File: backend/src/routes/education.ts
import { Router } from "express";
import { requireAuth, type AuthedRequest } from "../middleware/auth.js";
import { validateBody } from "../middleware/validate.js";
import { educationCreate, educationUpdate } from "../schemas/education.js";
import { getUserSupabaseClient } from "../supabaseClients.js";

export const educationRouter = Router();

educationRouter.use(requireAuth);

educationRouter.get("", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client.from("education").select("*").order("institution");
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }
  res.json(data);
});

educationRouter.post("", validateBody(educationCreate), async (req, res) => {
  const { token, userId } = req as AuthedRequest;
  const client = getUserSupabaseClient(token);
  const payload = { ...req.body, user_id: userId };
  const { data, error } = await client.from("education").insert(payload).select().single();
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }
  res.status(201).json(data);
});

educationRouter.get("/:educationId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("education")
    .select("*")
    .eq("id", req.params.educationId);
  if (error || !data?.length) {
    res.status(404).json({ detail: "Education entry not found" });
    return;
  }
  res.json(data[0]);
});

educationRouter.patch("/:educationId", validateBody(educationUpdate), async (req, res) => {
  if (Object.keys(req.body).length === 0) {
    res.status(400).json({ detail: "No fields to update" });
    return;
  }
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("education")
    .update(req.body)
    .eq("id", req.params.educationId)
    .select();
  if (error || !data?.length) {
    res.status(404).json({ detail: "Education entry not found" });
    return;
  }
  res.json(data[0]);
});

educationRouter.delete("/:educationId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  await client.from("education").delete().eq("id", req.params.educationId);
  res.status(204).send();
});

File: backend/app/routers/education.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from supabase import Client

from ..deps import get_current_user_id, get_user_supabase_client
from ..schemas import EducationCreate, EducationOut, EducationUpdate

router = APIRouter(prefix="/education", tags=["education"])


@router.get("", response_model=list[EducationOut])
def list_education(client: Client = Depends(get_user_supabase_client)) -> list[dict]:
    response = client.table("education").select("*").order("institution").execute()
    return response.data


@router.post("", response_model=EducationOut, status_code=status.HTTP_201_CREATED)
def create_education(
    education: EducationCreate,
    client: Client = Depends(get_user_supabase_client),
    user_id: str = Depends(get_current_user_id),
) -> dict:
    payload = education.model_dump(mode="json") | {"user_id": user_id}
    response = client.table("education").insert(payload).execute()
    return response.data[0]


@router.get("/{education_id}", response_model=EducationOut)
def get_education(
    education_id: UUID, client: Client = Depends(get_user_supabase_client)
) -> dict:
    response = (
        client.table("education").select("*").eq("id", str(education_id)).execute()
    )
    if not response.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Education entry not found")
    return response.data[0]


@router.patch("/{education_id}", response_model=EducationOut)
def update_education(
    education_id: UUID,
    education: EducationUpdate,
    client: Client = Depends(get_user_supabase_client),
) -> dict:
    payload = education.model_dump(mode="json", exclude_unset=True)
    if not payload:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No fields to update")
    response = (
        client.table("education")
        .update(payload)
        .eq("id", str(education_id))
        .execute()
    )
    if not response.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Education entry not found")
    return response.data[0]


@router.delete("/{education_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_education(
    education_id: UUID, client: Client = Depends(get_user_supabase_client)
) -> None:
    client.table("education").delete().eq("id", str(education_id)).execute()
